### PR TITLE
[read/write set analysis] Don't allow local or return paths in the fo…

### PR DESCRIPTION
**Pull request by sblackshear**
…otprint

Fixes the final crash issue required to analyze all Diem Framework modules.

- Add `Root::Formal` + use a `FunctionEnv` to decide whether to create a `Root::Local` or `Root::Formal`. This change is the bulk of the PR because we need to thread `FunctionEnv`'s to a lot of new places
- In `make_footprint(access_path)`, create an empty footprint if `access_path` is rooted in a local or return value. The `Root::Formal`/`Root::Local` split is required to detect this (since the call site of `make_footprint` can't get a `FunctionEnv`.
- Add minimal repro that triggered the crash in the old code.
- Add Diem framework as a smoke test.

### CODE REVIEW

File: language/move-prover/bytecode/src/access_path.rs [hunk 1]
```diff
@@ -257,7 +272,11 @@ impl AbsAddr {
 
 impl FootprintDomain for AbsAddr {
     fn make_footprint(ap: AccessPath) -> Option<Self> {
-        Some(AbsAddr::footprint(ap))
+        if !ap.root.is_return() && !ap.root.is_local() {
```
Comment (sblackshear) [inline language/move-prover/bytecode/src/access_path.rs:275]: Everything else in the PR is in service of implementing this check.

**Review by mengxu-fb — APPROVED** [review]
LGTM!

An alternative to stuffing `FunctionEnv` in function arguments is to put it in `ReadWriteSetState`, (like `current_function: &'env FunctionEnv<'env>`). I don't have a strong preference either way so feel free to land as it is.

**Comment by sblackshear** [comment]
> LGTM!
> 
> An alternative to stuffing `FunctionEnv` in function arguments is to put it in `ReadWriteSetState`, (like `current_function: &'env FunctionEnv<'env>`). I don't have a strong preference either way so feel free to land as it is.

This would also work. However, I did it this way like to keep a strong separation between mutable abstract state (e.g., `ReadWriteSetState`) and immutable supporting metadata (e.g., `FunctionEnv`). E.g., writing a join that accounts for `FunctionEnv` would be odd...

**Pull request merged**